File: src/differential_equations_exact_equations_easy.py
"""
differential_equations - exact_equations (easy)
Generated: 2026-02-22T05:21:35.471571
"""

from problem_utils import *

def generate():
    difficulty_choice = randint(1, 3)

    if difficulty_choice == 1:
        # Build from F = a*x^2/2 + b*x*y (M = a*x + b*y, N = b*x)
        a = randint(2, 8)
        b = randint(1, 6)

        M = a * x + b * y
        N = b * x

        dM_dy = diff(M, y)
        dN_dx = diff(N, x)

        f = integrate(M, x)
        g_prime = N - diff(f, y)
        g = integrate(g_prime, y)
        C_const = symbols('C')
        solution = Eq(f + g, C_const)

        question = f"Verify that $({latex(M)})dx + ({latex(N)})dy = 0$ is exact and find its implicit solution."

        sol_steps = steps(
            f"$M = {latex(M)}$, $N = {latex(N)}$",
            f"$\\frac{{\\partial M}}{{\\partial y}} = {latex(dM_dy)}$, $\\frac{{\\partial N}}{{\\partial x}} = {latex(dN_dx)}$ — equal, so exact",
            f"Integrate $M$ w.r.t. $x$: $F = {latex(f)} + g(y)$",
            f"Differentiate w.r.t. $y$: $\\frac{{\\partial F}}{{\\partial y}} = {latex(diff(f,y))} + g'(y) = N = {latex(N)}$",
            f"$g'(y) = {latex(g_prime)}$, $g(y) = {latex(g)}$",
            f"Implicit solution: ${latex(solution)}$"
        )

        return problem(
            question=question,
            answer=solution,
            difficulty=(1000, 1100),
            topic="differential_equations/exact_equations",
            solution=sol_steps,
            answer_type="equation"
        )

    elif difficulty_choice == 2:
        # F = a*x*y + b*x^2/2 + c*y^2/2  =>  M = a*y + b*x, N = a*x + c*y
        a = randint(1, 5)
        b = randint(1, 5)
        c = randint(1, 5)

        F = a * x * y + Rational(b, 2) * x**2 + Rational(c, 2) * y**2
        M = diff(F, x)
        N = diff(F, y)

        dM_dy = diff(M, y)
        dN_dx = diff(N, x)

        C_const = symbols('C')
        solution = Eq(F, C_const)

        question = f"Solve the exact differential equation $({latex(M)})dx + ({latex(N)})dy = 0$."

        integral_M = integrate(M, x)

        sol_steps = steps(
            f"$M = {latex(M)}$, $N = {latex(N)}$",
            f"$\\frac{{\\partial M}}{{\\partial y}} = {latex(dM_dy)} = \\frac{{\\partial N}}{{\\partial x}} = {latex(dN_dx)}$ ✓",
            f"Integrate $M$ w.r.t. $x$: ${latex(integral_M)} + g(y)$",
            f"$g'(y) = N - {latex(diff(integral_M, y))} = {latex(simplify(N - diff(integral_M, y)))}$",
            f"$g(y) = {latex(integrate(simplify(N - diff(integral_M, y)), y))}$",
            f"Solution: ${latex(solution)}$"
        )

        return problem(
            question=question,
            answer=solution,
            difficulty=(1100, 1200),
            topic="differential_equations/exact_equations",
            solution=sol_steps,
            answer_type="equation"
        )

    else:
        # Check exactness (yes/no)
        is_exact = choice([True, False])

        if is_exact:
            a = randint(1, 5)
            b = randint(1, 5)
            c = randint(1, 5)
            # Ensure exact: M_y = N_x
            M = a * x * y + b * x
            N = Rational(a, 2) * x**2 + c * y
            ans = True
        else:
            a = randint(2, 6)
            b = randint(1, 5)
            c = randint(1, 5)
            M = a * x + b * y
            N = c * x + b * y + 1  # M_y = b, N_x = c; if a != c then not exact
            while diff(M, y) == diff(N, x):
                c = randint(1, 8)
                N = c * x + b * y + 1
            ans = False

        dM_dy = diff(M, y)
        dN_dx = diff(N, x)

        question = f"Is the differential equation $({latex(M)})dx + ({latex(N)})dy = 0$ exact?"

        sol_steps = steps(
            f"$M = {latex(M)}$, $N = {latex(N)}$",
            f"$\\frac{{\\partial M}}{{\\partial y}} = {latex(dM_dy)}$",
            f"$\\frac{{\\partial N}}{{\\partial x}} = {latex(dN_dx)}$",
            f"Since $\\frac{{\\partial M}}{{\\partial y}} {'=' if ans else '\\neq'} \\frac{{\\partial N}}{{\\partial x}}$, the equation is {'exact' if ans else 'not exact'}."
        )

        return problem(
            question=question,
            answer=ans,
            difficulty=(1200, 1300),
            topic="differential_equations/exact_equations",
            solution=sol_steps,
            answer_type="boolean"
        )


emit(generate())

File: src/differential_equations_exact_equations_easy.jl
# differential_equations - exact_equations (easy)
# Generated: 2026-03-08T20:51:13.428660

include(joinpath(@__DIR__, "..", "julia", "src", "ProblemUtils.jl"))
using .ProblemUtils

@script x y begin
    set_topic!("differential_equations/exact_equations")
    
    # Generate exact differential equation at PRE-COMPETENT to BEGINNER level (700-1200)
    # Strategy: Create M(x,y)dx + N(x,y)dy = 0 where ∂M/∂y = ∂N/∂x
    
    problem_type = choice([:type1, :type2, :type3])
    
    if problem_type == :type1
        # Simple exact: (2x + y)dx + (x + 2y)dy = 0
        # ∂M/∂y = 1, ∂N/∂x = 1 ✓
        a = nonzero(-5, 5)
        b = nonzero(-5, 5)
        M = a*x + y
        N = x + b*y
        
        # Verify exactness
        dM_dy = diff(M, y)
        dN_dx = diff(N, x)
        
        # Solution: F(x,y) = C where ∂F/∂x = M
        # F = (a/2)x² + xy, then ∂F/∂y must equal N
        # ∂F/∂y = x, so we need N = x + (b/2)y² after integrating
        F = (a//2)*x^2 + x*y + (b//2)*y^2
        C = randint(-10, 10)
        
        question = "Verify that the differential equation \$$(tex(M))\\,dx + $(tex(N))\\,dy = 0\$ is exact, then solve it."
        answer = "$(tex(F)) = C"
        
        solution = steps(
            sol("M(x,y)", M),
            sol("N(x,y)", N),
            sol("\\frac{\\partial M}{\\partial y}", dM_dy),
            sol("\\frac{\\partial N}{\\partial x}", dN_dx),
            "Since \\(\\frac{\\partial M}{\\partial y} = \\frac{\\partial N}{\\partial x}\\), the equation is exact.",
            "Integrate M with respect to x: \\(F(x,y) = \\int $(tex(M))\\,dx = $(tex((a//2)*x^2 + x*y)) + g(y)\\)",
            "Differentiate with respect to y and set equal to N: \\(\\frac{\\partial F}{\\partial y} = $(tex(x)) + g'(y) = $(tex(N))\\)",
            "So \\(g'(y) = $(tex(b*y))\\), giving \\(g(y) = $(tex((b//2)*y^2))\\)",
            sol("Solution", F ~ "C")
        )
        
        difficulty = (700, 950)
        time = 180
        
    elseif problem_type == :type2
        # Polynomial exact: (ay² + bx)dx + (2axy + cy)dy = 0
        a = nonzero(-4, 4)
        b = nonzero(-6, 6)
        c = nonzero(-4, 4)
        
        M = a*y^2 + b*x
        N = 2*a*x*y + c*y
        
        dM_dy = diff(M, y)
        dN_dx = diff(N, x)
        
        # Solution: F = ax²y² + (b/2)x² + (c/2)y² = C
        F = a*x*y^2 + (b//2)*x^2 + (c//2)*y^2
        
        question = "Determine if the differential equation \$$(tex(M))\\,dx + $(tex(N))\\,dy = 0\$ is exact. If so, find the general solution."
        answer = "$(tex(F)) = C"
        
        solution = steps(
            sol("M(x,y)", M),
            sol("N(x,y)", N),
            sol("\\frac{\\partial M}{\\partial y}", dM_dy),
            sol("\\frac{\\partial N}{\\partial x}", dN_dx),
            "The equation is exact since the partial derivatives are equal.",
            "Integrate M: \\(F = \\int $(tex(M))\\,dx = $(tex(a*x*y^2 + (b//2)*x^2)) + g(y)\\)",
            "Differentiate and match to N: \\(\\frac{\\partial F}{\\partial y} = $(tex(2*a*x*y)) + g'(y) = $(tex(N))\\)",
            "Therefore \\(g'(y) = $(tex(c*y))\\), so \\(g(y) = $(tex((c//2)*y^2))\\)",
            sol("General solution", F ~ "C")
        )
        
        difficulty = (850, 1100)
        time = 210
        
    else  # type3
        # Simple with constants: (k₁y)dx + (k₁x + k₂)dy = 0
        k1 = nonzero(-6, 6)
        k2 = nonzero(-8, 8)
        
        M = k1*y
        N = k1*x + k2
        
        dM_dy = diff(M, y)
        dN_dx = diff(N, x)
        
        # F = k₁xy + k₂y = C
        F = k1*x*y + k2*y
        
        question = "Show that \$$(tex(M))\\,dx + $(tex(N))\\,dy = 0\$ is exact and find its solution."
        answer = "$(tex(F)) = C"
        
        solution = steps(
            sol("Given: M(x,y)", M),
            sol("Given: N(x,y)", N),
            sol("\\frac{\\partial M}{\\partial y}", dM_dy),
            sol("\\frac{\\partial N}{\\partial x}", dN_dx),
            "Exact since \\(\\frac{\\partial M}{\\partial y} = \\frac{\\partial N}{\\partial x}\\)",
            "Integrate M: \\(F = \\int $(tex(M))\\,dx = $(tex(k1*x*y)) + g(y)\\)",
            "Match \\(\\frac{\\partial F}{\\partial y}\\) to N: \\($(tex(k1*x)) + g'(y) = $(tex(N))\\)",
            "So \\(g'(y) = $(k2)\\), giving \\(g(y) = $(tex(k2*y))\\)",
            sol("Solution", F ~ "C")
        )
        
        difficulty = (700, 900)
        time = 150
    end
    
    problem(
        question=question,
        answer=answer,
        difficulty=difficulty,
        solution=solution,
        time=time,
        answer_type="expression",
        grading_mode="symbolic"
    )
end